File: section2/type_demo.py
# type_demo.py
# Demonstrates Python's type system: dynamic typing, strong typing, duck typing

# --- 1. Dynamic typing ---
# variables are not bound to a type - they can hold anything
x = 42
print(type(x))        # <class 'int'>
x = "now a string"
print(type(x))        # <class 'str'>
x = [1, 2, 3]
print(type(x))        # <class 'list'>

# --- 2. Strong typing ---
# Python refuses implicit coercion between incompatible types
a = 10
b = "5"
try:
    result = a + b    # TypeError - no silent coercion
except TypeError as e:
    print(f"TypeError: {e}")

# explicit conversion is required
result = a + int(b)
print(f"explicit conversion: {result}")  # 15

# string + string is fine, just no mixing
print("hello" + " world")   # "hello world"

# --- 3. Duck typing ---
# Python doesn't care about the declared type, only whether the object
# has the method/attribute being used
class Dog:
    def speak(self):
        return "Woof"

class Cat:
    def speak(self):
        return "Meow"

class Robot:
    def speak(self):
        return "Beep boop"

def make_it_speak(thing):
    # works on any object that has a speak() method
    print(thing.speak())

make_it_speak(Dog())
make_it_speak(Cat())
make_it_speak(Robot())

# --- 4. Type hints (Python 3.5+) - optional, not enforced at runtime ---
def add_numbers(a: int, b: int) -> int:
    return a + b

# hints are ignored at runtime - this still runs
print(add_numbers("hello", " world"))  # "hello world" - no error

File: section2/type_demo.cpp
// type_demo.cpp
// Demonstrates C++'s type system: static typing, type inference with auto,
// function overloading, and templates as compile-time structural typing

#include <iostream>
#include <string>
using namespace std;

// --- 1. Function overloading based on declared type ---
// the compiler picks the right version at compile time
int add(int a, int b) {
    cout << "[int add] ";
    return a + b;
}

double add(double a, double b) {
    cout << "[double add] ";
    return a + b;
}

// --- 2. Template - closest C++ gets to duck typing ---
// works for any type T that supports operator+
template<typename T>
T doubled(T x) {
    return x + x;
}

int main() {
    // --- 3. Static typing ---
    int a = 10;
    double b = 3.7;
    string s = "hello";

    // a = "text";  // uncomment to see: error: invalid conversion from 'const char*' to 'int'

    // --- 4. auto - compiler infers the type from the initializer ---
    auto c = 42;        // int
    auto d = 3.14;      // double
    auto e = string("world");  // string
    cout << "auto types: " << c << ", " << d << ", " << e << endl;

    // --- 5. Implicit narrowing conversion (may lose data) ---
    double f = 9.99;
    int g = (int)f;     // explicit cast required to make the intent clear
    cout << "9.99 cast to int: " << g << endl;  // 9 - truncated

    // mixing int and double - int is implicitly widened, not truncated
    double result = a + b;   // 10 + 3.7 = 13.7
    cout << "int + double: " << result << endl;

    // --- 6. Overloading demo ---
    cout << add(3, 4) << endl;        // calls int version -> 7
    cout << add(1.5, 2.5) << endl;    // calls double version -> 4.0

    // --- 7. Template demo ---
    cout << doubled(6) << endl;               // 12
    cout << doubled(string("hi")) << endl;    // "hihi"

    return 0;
}
